File: MongoBooks2/BooksControllerUtilities/DataClasses/LanguageAuthors.cs
namespace BooksControllerUtilities.DataClasses
{
    using System;
    using BooksCore.Books;

    public class LanguageAuthors
    {
        public string Name { get; set; }

        public int TotalBooksReadInLanguage { get; set; }

        public int TotalPagesReadInLanguage { get; set; }

        public int TotalBooksWorldWide { get; set; }

        public int TotalPagesWorldWide { get; set; }

        public float PercentageOfBooksRead { get; set; }

        public float PercentageOfPagesRead { get; set; }

        public Author[] Authors { get; set; }

        public LanguageAuthors()
        {

        }

        public LanguageAuthors(AuthorLanguage authorLanguage)
        {
            Name = authorLanguage.Language;
            TotalBooksReadInLanguage = authorLanguage.TotalBooksReadInLanguage;
            TotalPagesReadInLanguage = (int)authorLanguage.TotalPagesReadInLanguage;
            TotalBooksWorldWide = authorLanguage.TotalBooksWorldWide;
            TotalPagesWorldWide = (int)authorLanguage.TotalPagesWorldWide;
            PercentageOfBooksRead = (float)Math.Round(authorLanguage.PercentageOfBooksRead, 2);
            PercentageOfPagesRead = (float)Math.Round(authorLanguage.PercentageOfPagesRead, 2);

            Authors = new Author[authorLanguage.AuthorsInLanguage.Count];
            for (int i = 0; i < authorLanguage.AuthorsInLanguage.Count; i++)
            {
                Authors[i] = new Author(authorLanguage.AuthorsInLanguage[i]);
            }
        }
    }
}

File: MongoBooks2/HiGoogleCloud/ClientApp/src/app/Models/languageAuthors.ts
import { Author } from './author'

export interface ILanguageAuthors
{
  name: string;
  totalBooksReadInLanguage: number;
  totalPagesReadInLanguage: number;
  totalBooksWorldWide: number;
  totalPagesWorldWide: number;
  percentageOfBooksRead: number;
  percentageOfPagesRead: number;
  authors: Author[];
};

export class LanguageAuthors implements ILanguageAuthors
{
  constructor(
    public name: string = "",
    public totalBooksReadInLanguage: number = 0,
    public totalPagesReadInLanguage: number = 0,
    public totalBooksWorldWide: number = 0,
    public totalPagesWorldWide: number = 0,
    public percentageOfBooksRead: number = 0,
    public percentageOfPagesRead: number = 0,
    public authors: Author[] = new Array<Author>())
  {
  }

  public authorsNames: string;

  public static getAuthorsAsHtmlList(item: LanguageAuthors): string
  {
    var htmlList: string = "<ul style=\"list-style-type:none;\">";

    if (item.authors != null && item.authors.length > 0) {
      for (let author of item.authors) {
        htmlList += "<li>" + author.name + "</li>";
      }
    }

    htmlList += "</ul>";

    return htmlList;
  }
}
